editor/src/components/canvas/dom-sampler.tsx

```diff
@@ -0,0 +1,455 @@
+import {
+  collectDomElementMetadataForElement,
+  getAttributesComingFromStyleSheets,
+} from './dom-walker'
+import type { UiJsxCanvasContextData } from './ui-jsx-canvas'
+
+export function runDomSampler(
+  elementsToFocusOn: ElementsToRerender,
+  options: {
+    scale: number
+    selectedViews: Array<ElementPath>
+    metadataToUpdate: ElementInstanceMetadataMap
+    spyCollector: UiJsxCanvasContextData
+  },
+): { metadata: ElementInstanceMetadataMap; tree: ElementPathTrees } {
+  const canvasRootContainer = document.getElementById(CanvasContainerID)
+  if (canvasRootContainer == null) {
+    return {
+      metadata: options.metadataToUpdate,
+      tree: MetadataUtils.createElementPathTreeFromMetadata(options.metadataToUpdate), // TODO this should return a cached tree
+    }
+  }
+
+  const validPaths = getValidPathsFromCanvasContainer(canvasRootContainer)
+
+  const spyPaths = Object.keys(options.spyCollector.current.spyValues.metadata)
+  if (spyPaths.length === 0 && validPaths.length > 0) {
+    // if there are no spy elements, but we have valid paths, we probably encountered a transient case during a canvas runtime error
+    // the intended behavior of the editor during a runtime error is to preserve the existing metadata to avoid flickering
+    // so we return the existing metadata in that case
+    return {
+      metadata: options.metadataToUpdate,
+      tree: MetadataUtils.createElementPathTreeFromMetadata(options.metadataToUpdate), // TODO this should return a cached tree
+    }
+  }
+
+  let result: { metadata: ElementInstanceMetadataMap; tree: ElementPathTrees }
+  if (elementsToFocusOn == 'rerender-all-elements') {
+    result = collectMetadataForPaths(
+      canvasRootContainer,
+      validPaths,
+      validPaths,
+      {},
+      { ...options, spyPaths: spyPaths },
+    )
+  } else {
+    result = collectMetadataForPaths(
+      canvasRootContainer,
+      validPaths.filter((vp) =>
+        elementsToFocusOn.some((e) => {
+          const staticElement = EP.makeLastPartOfPathStatic(e)
+          return EP.pathsEqual(staticElement, vp) || EP.isParentOf(staticElement, vp)
+        }),
+      ),
+      validPaths,
+      { ...options.metadataToUpdate },
+      { ...options, spyPaths: spyPaths },
+    )
+  }
+
+  return result
+}
+
+function collectMetadataForPaths(
+  canvasRootContainer: HTMLElement,
+  pathsToCollect: Array<ElementPath>,
+  validPaths: Array<ElementPath>,
+  metadataToUpdate_MUTATE: ElementInstanceMetadataMap,
+  options: {
+    scale: number
+    selectedViews: Array<ElementPath>
+    spyCollector: UiJsxCanvasContextData
+    spyPaths: Array<string>
+  },
+): {
+  metadata: ElementInstanceMetadataMap
+  tree: ElementPathTrees
+} {
+  const containerRect = getCanvasRectangleFromElement(
+    canvasRootContainer,
+    options.scale,
+    'without-text-content',
+    'nearest-half',
+  )
+
+  const dynamicPathsToCollect: Array<ElementPath> = pathsToCollect
```

Comment:
You can probably save on building this array by doing the `forEach` on `pathsToCollect` instead.

Reply:
fixed this too, now all params are in an object